### PR TITLE
Binary format

### PR DESCRIPTION
**Pull request by peterjc**
This pull request addresses issue #2332

- [X] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [X] I have read the ``CONTRIBUTING.rst`` file, have run ``flake8`` locally, and
understand that AppVeyor and TravisCI will be used to confirm the Biopython unit
tests and style checks pass with these changes.

- [X] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

### CODE REVIEW

File: Bio/SeqRecord.py [hunk 1]
```diff
@@ -715,14 +716,26 @@ def __format__(self, format_spec):
         if format_spec in SeqIO._FormatToString:
             return SeqIO._FormatToString[format_spec](self)
 
-        # Harder case, make a temp handle instead
         if format_spec in SeqIO._BinaryFormats:
```
Comment (mdehoon) [inline Bio/SeqRecord.py]: A better way is to call `SeqIO.write(self, handle, format_spec)` with a `StringIO` handle regardless of whether `format_spec` is in `SeqIO._BinaryFormats`, and raise a `ValueError` if `SeqIO.write` raises a `TypeError`.

Comment (peterjc) [reply Bio/SeqRecord.py]: But we wouldn't know for sure that would be the only way to see a ``TypeError`` - we could introduce a new specific exception, but it seems overkill.

Comment (mdehoon) [reply Bio/SeqRecord.py]: Then let's leave this open until we drop Python2. With Python3, it's easy to check if a handle is in text or binary mode. Then `SeqIO.write` can raise a `TypeError` with the appropriate error message, which can then be checked by `SeqRecord.__format__`.

Comment (peterjc) [reply Bio/SeqRecord.py]: OK - there will be a lot of Python 2 cleanup to come 👍

Comment (peterjc) [reply Bio/SeqRecord.py]: Or did you mean leave this pull request open till then? I think it is good enough as it stands.

Comment (mdehoon) [reply Bio/SeqRecord.py]: Yes I meant to leave this pull request open until we drop python2. Then we should be able to find a better solution (one that does not involve checking in `SeqIO._BinaryFormats`).

Comment (mdehoon) [reply Bio/SeqRecord.py]: For example, you could generalize `SeqIO.write` to return a string (or bytes) if `handle` is `None`.

Comment (mdehoon) [reply Bio/SeqRecord.py]: Perhaps a better solution is to add a `stringify` function to `Bio.SeqIO` and its submodules that takes a `SeqRecord` and returns a string if the format is string-like, and raise a `ValueError` if the format is byte-like.

Comment (peterjc) [reply Bio/SeqRecord.py]: We have a dictionary of SeqRecord to string functions already, but in general this is not possible due to file formats with data-dependent headers/footers.

**Comment by peterjc** [comment]
Applied, as discussed above we can revisit this after dropping Python 2 (not long to wait now).

**Pull request merged**